Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,533 @@
+        assert(MemoryLayout<UInt64>.size == MemoryLayout<URingUserData>.size)
+        self.registrationID = UInt16(truncatingIfNeeded: registrationID.rawValue)
+        self.fileDescriptor = fileDescriptor
+        self.eventType = eventType
+        self.padding = 0
+    }
+
+    @inlinable init(rawValue: UInt64) {
+        let unpacked = IntegerBitPacking.unpackUInt32UInt16UInt8(rawValue)
+        self = .init(registrationID: SelectorRegistrationID(rawValue: UInt32(unpacked.1)),
+                     fileDescriptor: CInt(unpacked.0),
+                     eventType: CQEEventType(rawValue:unpacked.2)!)
+    }
+}
+
+extension UInt64 {
+    init(_ uringUserData: URingUserData) {
+        let fd = uringUserData.fileDescriptor
+        let eventType = uringUserData.eventType.rawValue
+        assert(fd >= 0, "\(fd) is not a valid file descriptor")
+        assert(eventType >= 0, "\(eventType) is not a valid eventType")
+
+        self = IntegerBitPacking.packUInt32UInt16UInt8(UInt32(fd),
+                                                       uringUserData.registrationID,
+                                                       eventType)
+    }
+}
+
+// These are the events returned up to the selector
+internal struct UringEvent {
+    var fd : CInt
+    var pollMask : UInt32
+    var registrationID : UInt16 // we just have the truncated lower 16 bits of the registrationID
+    var pollCancelled : Bool
+    init () {
+        self.fd = -1
+        self.pollMask = 0
+        self.registrationID = 0
+        self.pollCancelled = false
+    }
+}
+
+// This is the key we use for merging events in our internal hashtable
+struct fdEventKey: Hashable {
+    var fileDescriptor : CInt
+    var registrationID : UInt16 // we just have the truncated lower 16 bits of the registrationID
+
+    init(_ f: CInt, _ s : UInt16) {
+        self.fileDescriptor = f
+        self.registrationID = s
+    }
+}
+
+
+final internal class Uring {
+    internal static let POLLIN: CUnsignedInt = numericCast(CNIOLinux.POLLIN)
+    internal static let POLLOUT: CUnsignedInt = numericCast(CNIOLinux.POLLOUT)
+    internal static let POLLERR: CUnsignedInt = numericCast(CNIOLinux.POLLERR)
+    internal static let POLLRDHUP: CUnsignedInt = CNIOLinux_POLLRDHUP() // numericCast(CNIOLinux.POLLRDHUP) 
+    internal static let POLLHUP: CUnsignedInt = numericCast(CNIOLinux.POLLHUP)
+    internal static let POLLCANCEL: CUnsignedInt = 0xF0000000 // Poll cancelled, need to reregister for singleshot polls
+
+    private var ring = io_uring()
+    private let ringEntries: CUnsignedInt = 8192
+    private let cqeMaxCount : UInt32 = 8192 // this is the max chunk of CQE we take.
+
+    var cqes : UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>
+    var fdEvents = [fdEventKey : UInt32]() // fd, sequence_identifier : merged event_poll_return
+    var emptyCqe = io_uring_cqe()
+
+    internal func fd() -> Int32 {
```

Comment:
Any reason this isn't a computed property?

Reply:
Nope, just reflects that I've not Swiftified myself enough yet (WIP) ;-) Thanks, changing.